### PR TITLE
feat(web): check for low-probability exact + exact-key correction matches 📚 

### PR DESCRIPTION
**Pull request by jahorton**
One long-time pet peeve of mine when it comes to auto-correct:  it should **_never_** correct away from a perfectly-valid word of the language.  Even if `it's` is a more common word in English than `its`, I believe that `its` should be left in place.  (I find it quite the pain on iOS, as I've had to fight iOS to leave `its` before.)  

Even if we _were_ to auto-correct away, we should ensure that `its` is, at least, easily accessible on the banner as an option so that a user may at least _prevent_ auto-correction that way.  (Admittedly, iOS does present it... I probably need to adapt.)  That is... we need to detect exact matches + exact-key matches and ensure they're always visible, with priority.  This is something our engine currently doesn't do well, but thanks to #11869, we now have the perfect tool to remedy the issue.

Secondly... the `ModelCompositor.predict` method is _long_ and rather "monolithic".  It'd be nice to spin off as much "keep"-related handling as possible into its own method.

Unit tests for the new functionality will be added with #11944.  Ideally, this PR should not be merged until all PRs in the sequence up to and including #11944 are ready to go.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by mcdurdin** [comment]
> One long-time pet peeve of mine when it comes to auto-correct: it should **_never_** correct away from a perfectly-valid word of the language. Even if `it's` is a more common word in English than `its`, I believe that `its` should be left in place. (I find it quite the pain on iOS, as I've had to fight iOS to leave `its` before.)

I am tempted to make you feel better by saying "there, their, they're"...

Interestingly, I've seen other users who *love* this. `its` vs `it's` is impossible to get right without some grammatical awareness of course... but in the more general case, I wonder if this should be surfaced as an option for users?

**Comment by jahorton** [comment]
> > One long-time pet peeve of mine when it comes to auto-correct: it should **_never_** correct away from a perfectly-valid word of the language. Even if `it's` is a more common word in English than `its`, I believe that `its` should be left in place. (I find it quite the pain on iOS, as I've had to fight iOS to leave `its` before.)
> 
> I am tempted to make you feel better by saying "there, their, they're"...
> 
> Interestingly, I've seen other users who _love_ this. `its` vs `it's` is impossible to get right without some grammatical awareness of course... but in the more general case, I wonder if this should be surfaced as an option for users?

I was actually thinking something similar, making it a configurable option.

**Comment by mcdurdin** [comment]
> I was actually thinking something similar, making it a configurable option.

* #11931

File: common/models/templates/src/trie-model.ts [hunk 1]
```diff
@@ -453,7 +453,24 @@ class Trie {
...
     let searchKey = this.toKey(prefix);
     let rootTraversal = this.traverseFromRoot().child(searchKey);
-    return rootTraversal ? getSortedResults(rootTraversal) : [];
+
+    if(!rootTraversal) {
+      return [];
+    }
+
+    const directEntries = rootTraversal.entries;
+    // `Set` requires Chrome 38+, which is more recent than Chrome 35.
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:462]: If we are moving away from ES5 (#11881), does that bump our minimum version of Chrome?

File: common/models/templates/src/trie-model.ts [hunk 2]
```diff
@@ -453,7 +453,24 @@ class Trie {
...
+    const directSet: Record<string, string> = {};
+    for(let entry of directEntries) {
+      directSet[entry.text] = entry.text;
+    }
+
+    const bestEntries = getSortedResults(rootTraversal);
+    const deduplicated = bestEntries.filter((entry) => !directSet[entry.text])
+
+    // Any entries directly hosted on the current note should get full display
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:471]:
```suggestion
    // Any entries directly hosted on the current node should get full display
```

I think?

File: common/models/templates/src/trie-model.ts [hunk 3]
```diff
@@ -453,7 +453,24 @@ class Trie {
...
+
+    const directEntries = rootTraversal.entries;
+    // `Set` requires Chrome 38+, which is more recent than Chrome 35.
+    const directSet: Record<string, string> = {};
+    for(let entry of directEntries) {
+      directSet[entry.text] = entry.text;
+    }
+
+    const bestEntries = getSortedResults(rootTraversal);
+    const deduplicated = bestEntries.filter((entry) => !directSet[entry.text])
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:469]:
```suggestion
    const deduplicated = bestEntries.filter((entry) => !directSet[entry.text]);
```

File: common/models/templates/src/trie-model.ts [hunk 4]
```diff
@@ -453,7 +453,24 @@ class Trie {
...
-    return rootTraversal ? getSortedResults(rootTraversal) : [];
+
+    if(!rootTraversal) {
+      return [];
+    }
+
+    const directEntries = rootTraversal.entries;
+    // `Set` requires Chrome 38+, which is more recent than Chrome 35.
+    const directSet: Record<string, string> = {};
+    for(let entry of directEntries) {
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:464]:
```suggestion
    for(const entry of directEntries) {
```

File: common/models/templates/src/trie-model.ts [hunk 5]
```diff
@@ -453,7 +453,24 @@ class Trie {
   lookup(prefix: string): TextWithProbability[] {
     let searchKey = this.toKey(prefix);
     let rootTraversal = this.traverseFromRoot().child(searchKey);
```
Comment (mcdurdin) [inline common/models/templates/src/trie-model.ts:455]: These should be const as well?

```suggestion
    const searchKey = this.toKey(prefix);
    const rootTraversal = this.traverseFromRoot().child(searchKey);
```

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 6]
```diff
@@ -143,6 +164,11 @@ export default class ModelCompositor {
     let prefixTransform: Transform;
     let postContextState: correction.TrackedContextState = null;
 
+    let currentCasing: CasingForm = null;
+    if(lexicalModel.languageUsesCasing) {
+      currentCasing = this.detectCurrentCasing(postContext);
+    }
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/model-compositor.ts:170]:
```suggestion
    const currentCasing: CasingForm = lexicalModel.languageUsesCasing 
      ? this.detectCurrentCasing(postContext)
      : null;
```

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 7]
```diff
@@ -380,83 +404,83 @@ export default class ModelCompositor {
...
 
-          let intermediateKeep = models.transformToSuggestion(keepTransform, prob);
-          keepOption = this.toAnnotatedSuggestion(intermediateKeep, 'keep',  models.QuoteBehavior.noQuotes);
+      // Is the suggestion an exact match (or, "similar enough") to the
+      // actually-typed context?  If so, we wish to note this fact and to
+      // prioritize such a suggestion over suggestions that are not.
+      if(keyed(tuple.correction.sample) == keyedPrefix) {
+        if(predictedWord == truePrefix) {
+          tuple.matchLevel = SuggestionSimilarity.exact;
+          keepOption = this.toAnnotatedSuggestion(tuple.prediction.sample, 'keep',  models.QuoteBehavior.noQuotes);
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/model-compositor.ts:432]: It would be good to add a comment for this and the next 3 lines of code because it's not entirely obvious what's happening here, even with the comments on ll.426-428

File: common/web/lm-worker/src/main/model-compositor.ts [hunk 8]
```diff
@@ -569,10 +581,22 @@ export default class ModelCompositor {
   }
 
   private predictionAutoSelect(suggestionDistribution: CorrectionPredictionTuple[]) {
-    if(suggestionDistribution.length == 0) {
+    if(suggestionDistribution.length < 1) {
```
Comment (mcdurdin) [inline common/web/lm-worker/src/main/model-compositor.ts:584]: Why this change?

**Review by mcdurdin — APPROVED** [review]
I think this LGTM; I haven't reviewed the changes to the tests in detail as my flight is about to start descent and I wanted to get this in before that!

File: common/models/templates/src/trie-model.ts (hunk 1 above)
Comment (jahorton) [reply common/models/templates/src/trie-model.ts:462]: It does.  This PR's main content was written months ago, before we were looking at directly dropping it.  #11881 isn't yet merged, either.

**Comment by jahorton** [comment]
Noticed this while working on focused unit tests for the followup to #11940:  even with this PR in place, we're not currently auto-selecting something like `can't` with priority when the current context is `cant` - where there's no _exact_ context match, but there _is_ an exact-key match.  I'll want to fix that, whether it be within this PR or within a descendant.

The probability-ratio requirement should probably only apply to suggestions within the same "similarity tier".  If it's a lower tier, we should probably straight-up ignore its probability component for the sum used in the thresholding ratio.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.75-alpha](https://keyman.com/downloads/releases/alpha/18.0.75)
